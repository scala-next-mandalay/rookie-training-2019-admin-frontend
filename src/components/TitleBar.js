import React from 'react';
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/core/styles';
import { Box, AppBar, IconButton, Toolbar,Badge,Hidden } from '@material-ui/core';
import MenuIcon from '@material-ui/icons/Menu';
import AccountCircle from '@material-ui/icons/AccountCircle';
import MailIcon from '@material-ui/icons/Mail';
import NotificationsIcon from '@material-ui/icons/Notifications';
import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';
import Fade from '@material-ui/core/Fade';

const useStyles = makeStyles(theme => ({
  appBar: {
    zIndex: theme.zIndex.drawer + 1,
  },
  menuButton: {
    marginRight: theme.spacing(2),
    [theme.breakpoints.up('md')]: {
      display: 'none',
    },
  },
  rightIcon: {
    marginLeft: 'auto',
    marginRight: theme.spacing(1),
  }
}));

const TitleBar = ({handleDrawerToggle, changeAuthState, fetchAuthedUser, signOut, user, history}) => {
  const classes = useStyles();
  const [anchorEl, setAnchorEl] = React.useState(null);
  const open = Boolean(anchorEl);

  const isFirstRef = React.useRef(true);
  React.useEffect(() => {
    if (isFirstRef.current) {
      isFirstRef.current = false;
      fetchAuthedUser();
    }
  });
  
  const handleLogout = event => {
    event.preventDefault();
    signOut();
  };

  function handleClick(event) {
    setAnchorEl(event.currentTarget);
  }

  function handleClose() {
    setAnchorEl(null);
  }
  
  return (
    <AppBar position="fixed" className={classes.appBar}>
      <Toolbar>
        <IconButton
          color="inherit"
          aria-label="Open drawer"
          edge="start"
          onClick={handleDrawerToggle}
          disabled={(handleDrawerToggle === null)}
          className={classes.menuButton}
        >
          <MenuIcon />
        </IconButton>
        <Box fontSize="h6.fontSize" flex={1}>Management Console</Box>
        <Hidden only={["xs"]}>
        <IconButton aria-label="Show 4 new mails" color="inherit">
          <Badge badgeContent={4} color="secondary">
            <MailIcon />
          </Badge>
        </IconButton>
        <IconButton aria-label="Show 17 new notifications" color="inherit">
          <Badge badgeContent={17} color="secondary">
            <NotificationsIcon />
          </Badge>
        </IconButton>
        </Hidden>
        <IconButton
              edge="end"
              aria-label="Account of current user"
              color="inherit"
              aria-controls="fade-menu" 
              aria-haspopup="true" 
              onClick={handleClick}
        >
          <AccountCircle />
        </IconButton>
        <Menu
        id="fade-menu"
        anchorEl={anchorEl}
        keepMounted
        open={open}
        onClose={handleClose}
        TransitionComponent={Fade}
      >
        <MenuItem onClick={handleClose}>Profile</MenuItem>
        <MenuItem onClick={handleClose}>My account</MenuItem>
        <MenuItem onClick={handleLogout}>Logout</MenuItem>
      </Menu>
      </Toolbar>
    </AppBar>
  );
};

TitleBar.propTypes = {
  handleDrawerToggle: PropTypes.func,
  changeAuthState: PropTypes.func.isRequired,
  fetchAuthedUser: PropTypes.func.isRequired
};

TitleBar.defaultProps = {
  handleDrawerToggle: null,
};

export default TitleBar;