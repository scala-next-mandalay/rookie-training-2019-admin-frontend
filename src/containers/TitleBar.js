import { connect } from 'react-redux';
import TitleBar from '../components/TitleBar';
import { signOut } from '../modules/auth';
import {setLocale} from '../modules/locale';

export default connect(
  (state) => ({
    locale: state.locale.locale
  }),
  (dispatch) => ({
    signOut: () => dispatch(signOut()),
    setLocale: (value) => dispatch(setLocale(value)) 
  })

)(TitleBar);