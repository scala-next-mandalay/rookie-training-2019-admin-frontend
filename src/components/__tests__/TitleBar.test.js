/*global jest,expect,Event,MouseEvent*/
import React from 'react';
import 'jest-prop-type-error';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';
import renderer from 'react-test-renderer';
import Parent from '../Parent';
import TitleBar from '../TitleBar';
let container;

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.outerHTML = '<body></body>';
  container = null;
});

const fetchAuthedUser = jest.fn(() => {
  let user = [{ id: 1, emali: "a@gmail.com" }];
});

const changeAuthState = jest.fn(() => {
  let auth = true;
});

const signOut = jest.fn(()=>{
  let auth = false;
});
describe("Titalbar component", () => {
  it('matches the snapshot', () => {
    const TitalbarSnapshot= renderer.create(
      <Parent>
        <TitleBar
          fetchAuthedUser={fetchAuthedUser}
          changeAuthState={changeAuthState}
          signOut={signOut}/>
      </Parent>
    ).toJSON();
    expect(TitalbarSnapshot).toMatchSnapshot();
  });
});

describe("Teting titlebar", () => {
  it('Testing Signout button', () => {
    act(() => {
      ReactDOM.render((
        <Parent>
          <TitleBar
            fetchAuthedUser={fetchAuthedUser}
            changeAuthState={changeAuthState} 
            signOut={signOut}/>
        </Parent>
      ), container);
      
      const button = container.querySelectorAll('Button');
        act(() => {
          button[1].dispatchEvent(new MouseEvent('click', { bubbles: true }));
        });
      
      
      // cannot test becauseof  <Hidden /> but comment <Hidden/> the following testing work

      const ButtonArr = container.querySelectorAll('button');
      //console.log('HTML :',document.body.outerHTML);
        act(() => {
          ButtonArr[1].dispatchEvent(new MouseEvent('click', { bubbles: true }));
        });
      //console.log('HTML :',document.body.outerHTML);
      //const Button = container.querySelectorAll('Button');
      //console.log('AfterClicked :',Button);
      //console.log('Buttons :',Button.length)
        // act(() => {
        //   ButtonArr[2].dispatchEvent(new Event('click', { bubbles: true }));
        // });
      
        // expect(signOut).toHaveBeenCalled();
        
    });
   });
});