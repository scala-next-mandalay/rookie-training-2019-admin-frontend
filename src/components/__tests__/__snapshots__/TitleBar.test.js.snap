// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Titalbar component matches the snapshot 1`] = `
<header
  className="MuiPaper-root MuiPaper-elevation4 MuiAppBar-root MuiAppBar-positionFixed makeStyles-appBar-1 MuiAppBar-colorPrimary mui-fixed"
>
  <div
    className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
  >
    <button
      aria-label="Open drawer"
      className="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-2 MuiIconButton-colorInherit Mui-disabled MuiIconButton-edgeStart Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onClick={null}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
          />
        </svg>
      </span>
    </button>
    <div
      className="MuiBox-root MuiBox-root-67"
    >
      <span>
        Management Console
      </span>
    </div>
    <div
      className="MuiBox-root MuiBox-root-68"
    >
      <div
        className="MuiGrid-root MuiGrid-container"
      >
        <button
          aria-controls="lang-menu"
          aria-haspopup="true"
          aria-label="More"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</header>
`;
